Introduce sharding (and experimental auto-sharding)

brancz opened the pull request:
**What this PR does / why we need it**:

This PR introduces sharding capabilities for kube-state-metrics, as well as the experimental feature to have kube-state-metrics automatically detect sharding, by discovering its nominal position within a StatefulSet and using the number of replicas in the StatefulSet as the total number of shards. I added a further description of the mechanism to the documentation.

These features will allow kube-state-metrics to be easily auto-scaled with the horizontal-pod-autoscaler by using its latency of `/metrics` requests to be lower than a certain threshold, otherwise scaling it up, as the latency is directly related to the amount of objects a shard needs to handle and the more replicas in a setup, the lower the number of objects, therefore a lower latency, as each shard handles 1/nth of objects in the cluster.

kube-state-metrics is well optimized, but in order to scale it infinitely and on commodity hardware it is necessary to shard the work it needs to do.

Note, I do not intend to ship these features in the already started release series of 1.5.x of kube-state-metrics, but rather release following it.

@andyxning @zouyee @mxinden @s-urbaniak @squat @metalmatze

mxinden commented on README.md:
```diff
@@ -132,6 +139,23 @@ These numbers are based on [scalability tests](https://github.com/kubernetes/kub
+
+In order to horizontally scale kube-state-metrics some automated sharding capabilities have been implemented. It is configured with the following flags:
+
+* `--shard`
```
Maybe worth mentioning whether these are zero indexed?

mxinden commented on README.md:
```diff
@@ -132,6 +139,23 @@ These numbers are based on [scalability tests](https://github.com/kubernetes/kub
+
+There is also an experimental feature, that allows kube-state-metrics to auto discover its nominal position if it is deployed in a StatefulSet, in order to automatically configure sharding. This is an experimental feature and may be broken or removed without notice.
+
+To enable automated sharding kube-state-metrics must be run by a `StatefulSet` and the pod names and namespace must be handed to the kube-state-metrics process via the `--pod` and `--pod-namespace` flags.
```
Instead of kube-state-metrics reaching out to the API-Server to get its index in the total amount of shards, why not derive it from the pod name retrieved via an environment variable?

brancz replied:
kube-state-metrics needs to know the total number of replicas, that’s what it requests the statefulset for. Detecting it’s nominal within it is done as you said.

mxinden replied:
Right. Sorry about that. And this total # can not be templated from the statefulset status as an environment variable?

brancz replied:
The downward API only works for data within the pod object, not higher level objects.

mxinden commented on pkg/sharding/listwatch.go:
```diff
@@ -0,0 +1,88 @@
+package sharding
+
+import (
+	"crypto/md5"
```
Why use a `crypto` hash function in this context? Wouldn't one of https://golang.org/pkg/hash/ be faster?

mxinden commented on pkg/sharding/listwatch.go:
```diff
@@ -0,0 +1,88 @@
+func NewShardedListWatch(shard, totalShards uint64, lw cache.ListerWatcher) cache.ListerWatcher {
+	// This is an "optimization" as this configuration means no sharding is to
+	// be performed.
+	if shard == 0 && totalShards == 1 {
```
I guess if sharding is not set in `main.go`, these two would default to zero. Don't we want to just do `return lw` in that case? Hope I am not missing something.

brancz replied:
shard defaults to 0 and total shards defaults to 1

mxinden commented on pkg/sharding/listwatch.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// sum64 sums the md5 hash to an uint64.
+func sum64(hash [md5.Size]byte) uint64 {
```
If we use the golang/hash package, we could use the `Sum64` interface function instead, right?

```golang
type Hash64 interface {
        Hash
        Sum64() uint64
}
```

s-urbaniak replied:
Choosing the right hash function, woah ;-) I am literally not an expert, but I think the focus is here on good key distribution and speed. When we talk about "speed", the amount of data is not super critical here I guess as there are not that many objects to be calculated per second.

While looking at https://github.com/golang/groupcache/blob/c65c006176ff7ff98bb916961c7abbc6b0afc0aa/consistenthash/consistenthash.go#L42, crc32 might be fine, but https://medium.com/@dgryski/consistent-hashing-algorithmic-tradeoffs-ef6b8e2fcae8 points out some pros/cons.

brancz replied:
I chose this exact hashing mechanism as it's been well proven in the Prometheus `hashmod` system.

mxinden commented on compare-sharding.sh:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/bash
```
I would prefer this script not to be on the root of the repo. How about the `tests` directory instead?

zouyee commented on main.go:
```diff
@@ -272,3 +287,37 @@ func (m *metricHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	owners := p.GetOwnerReferences()
+	for _, o := range owners {
+		if o.APIVersion == "apps/v1" && o.Kind == "StatefulSet" {
```
wether we need to detect pod actual Controller?

brancz replied:
Not strictly necessary I think, it's a heuristic anyways, as theoretically there could be any number of statefulsets in the ownerreferences, and we just pick the first occurrence. I'll add it anyways though.

s-urbaniak commented on main.go:
```diff
@@ -272,3 +287,37 @@ func (m *metricHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	owners := p.GetOwnerReferences()
+	for _, o := range owners {
+		if o.APIVersion == "apps/v1" && o.Kind == "StatefulSet" {
```
stylistic nit to reduce the amount of indentation:
```
if o.APIVersion != "apps/v1" || o.Kind != "StatefulSet" {
  continue
}
// unindented logic
```

s-urbaniak commented on main.go:
```diff
@@ -272,3 +287,37 @@ func (m *metricHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	glog.Infof("Auto detecting shard.")
+	p, err := kubeClient.CoreV1().Pods(namespaceName).Get(podName, metav1.GetOptions{})
+	if err != nil {
+		return 0, 0, err
```
nit: let's add some context to indicate where we are, i.e. `fmt.Errorf("error detecting shard: %v", err)`

s-urbaniak commented on main.go:
```diff
@@ -272,3 +287,37 @@ func (m *metricHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		if o.APIVersion == "apps/v1" && o.Kind == "StatefulSet" {
+			ss, err := kubeClient.AppsV1().StatefulSets(namespaceName).Get(o.Name, metav1.GetOptions{})
+			if err != nil {
+				return 0, 0, err
```
same here, i.e.`fmt.Errorf("error detecting shard statefulset: %v", err)`

s-urbaniak commented on main.go:
```diff
@@ -272,3 +287,37 @@ func (m *metricHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+			nominalString := strings.TrimLeft(ss.Name+"-", p.Name)
+			nominal, err := strconv.Atoi(nominalString)
+			if err != nil {
+				return 0, 0, err
```
same here, i.e. `fmt.Errorf("error detecting shard index: %v", err)`

brancz commented:
I addressed the comments, but this also needs a reconcile loop, in cases where the statefulset is resized. I'd say we address that in a follow up as it requires a larger refactoring of the main.go in order to allow reloads of internals (including proper handling and passing of context).

mxinden commented:
Sounds good @brancz.

tariq1890 commented on kubernetes/autosharding/kube-state-metrics-statefulset.yaml:
```diff
@@ -0,0 +1,44 @@
+      serviceAccountName: kube-state-metrics
+      containers:
+      - name: kube-state-metrics
+        image: quay.io/coreos/kube-state-metrics:v1.4.0
```
```suggestion
        image: quay.io/coreos/kube-state-metrics:v1.5.0
```

tariq1890 replied:
updated my review comment.

tariq1890 commented on main.go:
```diff
@@ -28,14 +28,20 @@ import (
 	"strings"
+	"sync"
 
+	"github.com/golang/glog"
```
We should be using `klog`.

tariq1890 commented on main.go:
```diff
@@ -243,11 +253,105 @@ func serveMetrics(collectors []*coll.Collector, host string, port int, enableGZI
+		glog.Infof("sharding disabled")
+	}
+	if totalShards > 1 {
+		glog.Infof("configuring sharding of this instance to be shard %d out of %d", shard, totalShards)
```
```suggestion
		klog.Infof("configuring sharding of this instance to be shard %d out of %d", shard, totalShards)
```

tariq1890 commented on main.go:
```diff
@@ -243,11 +253,105 @@ func serveMetrics(collectors []*coll.Collector, host string, port int, enableGZI
+
+			shard, totalShards, err := shardingSettingsFromStatefulSet(ss, m.opts.Pod)
+			if err != nil {
+				glog.Errorf("detect sharding settings from StatefulSet: %v", err)
```
```suggestion
				klog.Errorf("detect sharding settings from StatefulSet: %v", err)
```

tariq1890 commented on main.go:
```diff
@@ -243,11 +253,105 @@ func serveMetrics(collectors []*coll.Collector, host string, port int, enableGZI
+		return ctx.Err()
+	}
+
+	glog.Infof("Auto detecting sharding settings.")
```
```suggestion
	klog.Infof("Auto detecting sharding settings.")
```

tariq1890 commented on main.go:
```diff
@@ -243,11 +253,105 @@ func serveMetrics(collectors []*coll.Collector, host string, port int, enableGZI
+
+			shard, totalShards, err := shardingSettingsFromStatefulSet(ss, m.opts.Pod)
+			if err != nil {
+				glog.Errorf("detect sharding settings from StatefulSet: %v", err)
```
```suggestion
				klog.Errorf("detect sharding settings from StatefulSet: %v", err)
```

tariq1890 commented on main.go:
```diff
@@ -276,3 +380,42 @@ func (m *metricHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		return ss, nil
+	}
+
+	return nil, fmt.Errorf("no suitable statefulset found for auto detecting sharding for Pod %s/%s", namespaceName, podName)
```
```suggestion
	return nil, errors.Errorf("no suitable statefulset found for auto detecting sharding for Pod %s/%s", namespaceName, podName)
```

tariq1890 commented on pkg/options/options.go:
```diff
@@ -82,6 +86,13 @@ func (o *Options) AddFlags() {
+	o.flags.Uint64Var(&o.Shard, "shard", uint64(0), "The instances shard nominal (zero indexed) within the total number of shards. (default 0)")
+	o.flags.Uint64Var(&o.TotalShards, "total-shards", uint64(1), "The total number of shards. Sharding is disabled when total shards is set to 1.")
+
+	autoshardingNotice := "When set, it is expected that --pod and --pod-namespace are both set. This is used for auto-detecting sharding. If set this has preference over statically configured sharding. This is exprimental, it may be removed without notice."
```
```suggestion
	autoshardingNotice := "When set, it is expected that --pod and --pod-namespace are both set. This is used for auto-detecting sharding. If set, this has preference over statically configured sharding. This is experimental, it may be removed without notice."
```

brancz commented:
I have yet to address all comments here, but autosharding including re-sharding seems to work now! :tada: I have yet to also add e2e tests for this. But it's progressing.

brancz commented:
I finally figured out why the benchmark seemed so much worse with these patches even though there is no change in the http serving parts: the go compiler can do unfair optimizations with the code on master. Once we change that, the result becomes what I would have expected. I opened #784 to address this.

tariq1890 commented on pkg/metricshandler/metrics_handler.go:
```diff
@@ -0,0 +1,244 @@
+	})
+	go i.Run(ctx.Done())
+	if !cache.WaitForCacheSync(ctx.Done(), i.HasSynced) {
+		return fmt.Errorf("waiting for informer cache to sync failed")
```
```suggestion
		return errors.New("waiting for informer cache to sync failed")
```

tariq1890 commented on pkg/metricshandler/metrics_handler.go:
```diff
@@ -0,0 +1,244 @@
+	i.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(o interface{}) {
+			ss := o.(*appsv1.StatefulSet)
+			if ss.Name != statefulSetName {
```
Let's address the code duplication in AddFunc and UpdateFunc.

brancz replied:
I don't think this is worth it, and the "skip" logic is slightly different for add and update, as in update we also get to compare to the previous object. I think this duplication is ok in its current extend.

tariq1890 commented on pkg/metricshandler/metrics_handler.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+func detectStatefulSet(kubeClient kubernetes.Interface, podName, namespaceName string) (*appsv1.StatefulSet, error) {
+	p, err := kubeClient.CoreV1().Pods(namespaceName).Get(podName, metav1.GetOptions{})
```
Is it possible to use a field selector within `metav1.GetOptions`so that pods could be pre-filtered by client-go?

brancz replied:
we already select exact pods, and not list

brancz commented:
That minikube failure does not look good. I think we'll be seeing that on all builds.

brancz commented:
Yep, we're seeing it everywhere. This is a master build with the same failure: https://travis-ci.org/kubernetes/kube-state-metrics/jobs/562200973

tariq1890 commented on pkg/options/options.go:
```diff
@@ -37,6 +37,10 @@ type Options struct {
+	Shard                                int32
+	TotalShards                          int
+	Pod                                  string
+	Namespace                            string
```
Let's rename this to `PodNamespace`. Since we already have `Namespaces`, `PodNamespace` would be a bit more clearer to the uninformed reader.

tariq1890 commented on docs/cli-arguments.md:
```diff
@@ -42,12 +42,16 @@ Usage of ./kube-state-metrics:
       --metric-blacklist string                     Comma-separated list of metrics not to be enabled. This list comprises of exact metric names and/or regex patterns. The whitelist and blacklist are mutually exclusive.
       --metric-whitelist string                     Comma-separated list of metrics to be exposed. This list comprises of exact metric names and/or regex patterns. The whitelist and blacklist are mutually exclusive.
       --namespace string                            Comma-separated list of namespaces to be enabled. Defaults to ""
+      --pod string                                  Name of the pod. When set, it is expected that --pod and --pod-namespace are both set. This is used for auto-detecting sharding. If set, this has preference over statically configured sharding. This is experimental, it may be removed without notice.
```
Name of the pod isn't too clear. We need to mention that this is the pod that corresponds to the ksm binary itself.

Thoughts on this - "Name of the pod that contains the kube-state-metrics container"?

brancz replied:
Yes I think that's a good idea. I also adapted the comment to additionally say that people are most likely going to need to do this via the downward api.

tariq1890 commented on docs/cli-arguments.md:
```diff
@@ -42,12 +42,16 @@ Usage of ./kube-state-metrics:
       --metric-whitelist string                     Comma-separated list of metrics to be exposed. This list comprises of exact metric names and/or regex patterns. The whitelist and blacklist are mutually exclusive.
       --namespace string                            Comma-separated list of namespaces to be enabled. Defaults to ""
+      --pod string                                  Name of the pod. When set, it is expected that --pod and --pod-namespace are both set. This is used for auto-detecting sharding. If set, this has preference over statically configured sharding. This is experimental, it may be removed without notice.
+      --pod-namespace string                        Name of the namespace. When set, it is expected that --pod and --pod-namespace are both set. This is used for auto-detecting sharding. If set, this has preference over statically configured sharding. This is experimental, it may be removed without notice.
```
Isn't it safe to assume the default pod and pod namespaces names to be - `kube-state-metrics` and `kube-system` respectively? 

If so, we can remove the requirement for both pod name and pod namespace to be set.

brancz replied:
We can't assume that I think and I'd prefer to have to explicitly set this instead of a default. I feel a default would be more error prone. I prefer to handle this via documented manifests.

tariq1890 commented on pkg/metricshandler/metrics_handler.go:
```diff
@@ -0,0 +1,243 @@
+	klog.Infof("Auto detecting sharding settings.")
+	ss, err := detectStatefulSet(m.kubeClient, m.opts.Pod, m.opts.Namespace)
+	if err != nil {
+		return errors.Wrap(err, "detect StatefulSet")
```
I see several instances of these error message styles. Can that be fixed everywhere?
```suggestion
		return errors.Wrap(err, "error when attempting to detect StatefulSet")
```

brancz replied:
The way it is is best practice. We're essentially building a stack trace.

tariq1890 replied:
Sure.

tariq1890 commented on pkg/metricshandler/metrics_handler.go:
```diff
@@ -0,0 +1,243 @@
+func shardingSettingsFromStatefulSet(ss *appsv1.StatefulSet, podName string) (nominal int32, totalReplicas int, err error) {
+	nominal, err = detectNominalFromPod(ss.Name, podName)
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "detecting Pod nominal")
```
Do we need to wrap the error here?

tariq1890 commented on tests/compare-sharding.sh:
```diff
@@ -0,0 +1,24 @@
+RESULT_SHARDED_UNSORTED="$(mktemp)"
+RESULT_SHARDED="$(mktemp)"
+
+curl localhost:8080/metrics | grep -v "^#" | sort | tee "${RESULT_UNSHARDED}"
```
can we have a ports array and loop through that?

brancz replied:
I don't think we can, as they are slightly different each time. Also I like the explicitness, and it's just a script for manual verification.

tariq1890 replied:
I can agree with that.

tariq1890 commented on tests/compare-sharding.sh:
```diff
@@ -0,0 +1,24 @@
+
+trap 'kill 0' SIGTERM
+
+kubectl -n kube-system port-forward deployment/kube-state-metrics 8080:8080 &
```
Let's move `kube-system` and `kube-state-metrics` to a variable. This way, we are able to correlate the `pod_name` and `pod_namespace` parameters specified in the README with those mentioned in this test.

brancz replied:
hmm .. how strongly do you feel about this, I quite like the explicitness, and feel introducing variables would make this harder to understand

tariq1890 replied:
I can see that. Works for me.

tariq1890 commented on tests/compare-sharding.sh:
```diff
@@ -0,0 +1,24 @@
+curl localhost:8084/metrics | grep -v "^#" | tee -a "${RESULT_SHARDED_UNSORTED}"
+curl localhost:8086/metrics | grep -v "^#" | tee -a "${RESULT_SHARDED_UNSORTED}"
+
+sort "$RESULT_SHARDED_UNSORTED" | tee "${RESULT_SHARDED}"
```
```suggestion
sort "${RESULT_SHARDED_UNSORTED}" | tee "${RESULT_SHARDED}"
```

tariq1890 commented on tests/compare-sharding.sh:
```diff
@@ -0,0 +1,24 @@
+
+sort "$RESULT_SHARDED_UNSORTED" | tee "${RESULT_SHARDED}"
+
+diff <(echo "$RESULT_UNSHARDED") <(echo "$RESULT_SHARDED")
```
```suggestion
diff <(echo "${RESULT_UNSHARDED}") <(echo "${RESULT_SHARDED}")
```

tariq1890 commented:
@brancz Can you rebase?

@LiliC Requesting your eyes on this as well.

LiliC commented on pkg/metricshandler/metrics_handler.go:
```diff
@@ -0,0 +1,243 @@
+func (m *MetricsHandler) Run(ctx context.Context) error {
+	autoSharding := len(m.opts.Pod) > 0 && len(m.opts.Namespace) > 0
+
+	klog.Infof("autosharding enabled: %t; pod=%v; pod_namespace=%v", autoSharding, m.opts.Pod, m.opts.Namespace)
```
Hmm not sure I like that this log is always logging. Also is it true that autosharding is enabled, how do we know that? From the options warning, it says you need to enable it? Can you elaborate, thanks! 

Nits: Also can we do the following:
- set this log level to debug
- default values for those fields are empty strings, meaning this msg will not make much sense.

brancz replied:
It's only logged once at startup and is very valuable information, I'd prefer to keep it at info level. Otherwise I split the log line into two (one enabled, one disabled) to make it less confusing.

LiliC commented on pkg/metricshandler/metrics_handler.go:
```diff
@@ -0,0 +1,243 @@
+	autoSharding := len(m.opts.Pod) > 0 && len(m.opts.Namespace) > 0
+
+	klog.Infof("autosharding enabled: %t; pod=%v; pod_namespace=%v", autoSharding, m.opts.Pod, m.opts.Namespace)
+	if !autoSharding {
```
Hmm should this not be before the log?

brancz replied:
changed the log line to be less confusing

LiliC commented on pkg/options/options.go:
```diff
@@ -84,6 +88,13 @@ func (o *Options) AddFlags() {
+	autoshardingNotice := "When set, it is expected that --pod and --pod-namespace are both set. This is used for auto-detecting sharding. If set, this has preference over statically configured sharding. This is experimental, it may be removed without notice."
+
+	o.flags.StringVar(&o.Pod, "pod", "", "Name of the pod. "+autoshardingNotice)
+	o.flags.StringVar(&o.Namespace, "pod-namespace", "", "Name of the namespace. "+autoshardingNotice)
```
Will this maybe `""` default to `default` namespace, because k8s? 🤔 :)

brancz replied:
It doesn’t, that’s kubectl Logic doing that. Here we’re referencing a concrete object so this isn’t correct.

brancz commented on kubernetes/autosharding/kube-state-metrics-cluster-role.yaml:
```diff
@@ -0,0 +1,60 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
The autosharding manifests use a statefulset instead of deployment to allow the identification of a nominal within a statefulset (as in the "self replica" out of the "total replicas"). As  well as the permissions for the pods to discover their owning statefulset.

brancz replied:
Let me know if it still needs to be clarified elsewhere. I added a comment to the readme referencing the manifests.

brancz replied:
Discussed out of band: we separated the "regular" deployment option and autosharding into separate directories.

brancz commented on kubernetes/autosharding/kube-state-metrics-cluster-role-binding.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: rbac.authorization.k8s.io/v1
+# kubernetes versions before 1.8.0 should use rbac.authorization.k8s.io/v1beta1
```
removed :)

brancz commented on pkg/metricshandler/metrics_handler.go:
```diff
@@ -0,0 +1,243 @@
+	curTotalShards int
+}
+
+func New(opts *options.Options, kubeClient kubernetes.Interface, storeBuilder *store.Builder, enableGZIPEncoding bool) *MetricsHandler {
```
Done.

brancz commented:
I believe I've fixed all comments and or replied to them and CI is green :green_apple: . Please take another look. :slightly_smiling_face:

tariq1890 commented on pkg/metricshandler/metrics_handler.go:
```diff
@@ -0,0 +1,257 @@
+	nominalString := strings.TrimPrefix(podName, statefulSetName+"-")
+	nominal, err := strconv.Atoi(nominalString)
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to detect shard index for Pod %s of StatefulSet %s, parsed %v", podName, statefulSetName, nominalString)
```
Let's use the string directive here.
```suggestion
		return 0, errors.Wrapf(err, "failed to detect shard index for Pod %s of StatefulSet %s, parsed %s", podName, statefulSetName, nominalString)
```

tariq1890 commented:
@brancz I think we should treat the moving of k8s manifests to a different folder as a breaking change. I would prefer if we kept the current directory setup as-is and created a separate folder for the autosharding scenarios.

I know that there are quite a few blogs and books that rely on the assumption that k8s manifests are located in the `kubernetes` folder.

brancz commented:
I think you're right. I reverted the regular deployment manifests to stay in the same place. PTAL :slightly_smiling_face:

tariq1890 commented on kubernetes/autosharding/kube-state-metrics-statefulset.yaml:
```diff
@@ -0,0 +1,47 @@
+          containerPort: 8080
+        - name: telemetry
+          containerPort: 8081
+        readinessProbe:
```
Can we add `livenessProbe` too?

tariq1890 commented on kubernetes/autosharding/kube-state-metrics-statefulset.yaml:
```diff
@@ -0,0 +1,53 @@
+          timeoutSeconds: 5
+        readinessProbe:
+          httpGet:
+            path: /healthz
```
the readiness should hit the base correct?
```suggestion
            path: /
```

tariq1890 commented:
/lgtm

Thanks @brancz! Great to have this merged :).
